converter/geant4/gdml/builder.py

```diff
@@ -0,0 +1,93 @@
+from typing import Optional, Set
+# skipcq: BAN-B405
+import xml.etree.ElementTree as ET
+from defusedxml.minidom import parseString
+
+from converter.geant4 import utils
+from .materials import collect, emit
+from .structure import emit_structure
+
+
+def generate_gdml_entry_point(world_json: Optional[dict]) -> str:
+    """Public entry: generate GDML string for provided world or an empty world."""
+    if world_json:
+        return _generate_gdml(world_json)
+    return _generate_empty()
+
+
+def _generate_gdml(world: dict) -> str:
+    """Generate GDML from geometry node"""
+    gdml_root = ET.Element("gdml", {
+        "xmlns:xsi": "http://www.w3.org/2001/XMLSchema-instance",
+        "xsi:noNamespaceSchemaLocation":
+            "http://cern.ch/service-spi/app/releases/GDML/schema/gdml.xsd",
+    })
+
+    ET.SubElement(gdml_root, "define")
+
+    used_materials: Set[str] = set()
+    collect(world, used_materials)
+
+    materials_xml = ET.SubElement(gdml_root, "materials")
+    emit(materials_xml, used_materials)
+
+    solids_xml = ET.SubElement(gdml_root, "solids")
+    structure_xml = ET.SubElement(gdml_root, "structure")
+
+    counters = {"solid": {}, "logic": {}, "phys": {}}
+
+    emit_structure(
+        node=world,
+        solids_xml=solids_xml,
+        structure_xml=structure_xml,
+        counters=counters,
+    )
+
+    setup = ET.SubElement(gdml_root, "setup", {
+        "name": "Default",
+        "version": "1.0",
+    })
+    ET.SubElement(setup, "world", {
+        "ref": utils.choose_logic_name(world),
+    })
+
+    return _prettify_xml(gdml_root)
```

Comment:
as a separate issue (not in this PR), we can think later of validation of the genenerated GDML string agains GDML schema, using i.e. lxml library.